### PR TITLE
Implement DerefMut for String

### PR DESCRIPTION
**Pull request by SimonSapin**
See https://github.com/rust-lang/rfcs/issues/1157

### CODE REVIEW

**Comment by SimonSapin** [comment]
Is it also time to look at either deprecating or stabilizing `OwnedAsciiExt` and `make_ascii_*case`?

**Comment by bluss** [comment]
Those casing ascii methods are pretty broken in unicode context, but I guess we could allow arbitrary ascii -> ascii mappings. Since a grapheme can be multiple codepoints, this also edits / transforms graphemes that use at least one ascii codepoint, for example é (depending on normalization).

**Comment by SimonSapin** [comment]
@bluss Agreed, the `AsciiExt` methods are for very narrow use case like CSS property names (which are entirely within ASCII), which is why they’re not inherent or in the prelude. We also have `to_lowercase` and `to_uppercase` on `char` and `str` that do use the full Unicode mappings.

**Comment by alexcrichton** [comment]
I would be more uncomfortable with stabilizing `make_ascii_{upper,lower}case` (primarily in terms of naming), but I'm ok with this. I'm going to tag this with T-libs and I-needs-decision to ensure it gets some visibility.

**Comment by SimonSapin** [comment]
OK. `make_ascii_*case` is a motivation for this change, but changing or stabilising would be a separate decision.

**Comment by Gankro** [comment]
I say we just land DerefMut and call it a day. Other stuff is easy to impl outside std.

**Comment by bluss** [comment]
I don't want to complicate this issue, but we might want to add slicing of `&mut str` and `.split_at_mut(mid)`, the latter being an important crutch for slices already to avoid needing unsafe sections.

**Comment by SimonSapin** [comment]
Added a commit:

>    Implement IndexMut for String and str.
    
>    ... matching the existing Index impls.
    There is no reason not to if String implement DerefMut.
    
>    The code removed in `src/librustc/middle/effect.rs` was added in #9750
    to prevent things like `s[0] = 0x80` where `s: String`,
    but I belive became unnecessary when the Index(Mut) traits were introduced.

**Comment by aturon** [comment]
I'm :+1: on this change in general, but are we OK landing this insta-stable? (Since it's just filling out APIs according to vetted/stable convention, seems fine to me, but I wanted to check.)

**Comment by SimonSapin** [comment]
`str::split_at_mut` is unstable (like the existing `str::split_at`) but yes, the rest (`DerefMut` and `IndexMut` impls as well as `str::slice_mut_unchecked`) is insta-stable in the current form of this PR. I don’t mind changing it.

**Comment by aturon** [comment]
Hm, I was also forgetting that implementations of stable traits are instantly stable regardless. I'm happy with the PR as it stands.

**Comment by alexcrichton** [comment]
This PR is now entering its final comment period. There was some discussion at triage today about this and there was some hesitation about the power that `DerefMut` is adding, so we decided to FCP it instead of merging it immediately!

**Comment by SimonSapin** [comment]
After seeing #9750 again I’ve been pondering on whether this PR introduces a way to break the UTF-8 invariant in safe code. I believe it does not:

* With `x: &mut T`, `*T` can only be assigned to directly when `T` is sized. Since `str: !Sized`, `&mut str` can not be assigned to directly.
* The `IndexMut` impls in this PR all have `Target=str`, so from a `&mut str` they can only give another `&mut str`.
* Although `Index<usize, Target=u8> for str` would make sense (and did sort of exist in Rust once upon a time: #9750) but is not implemented. The corresponding `IndexMut` would allow string mutation, but it’s not implemented either.
* The above leaves as the only way to mutate `&mut str` from safe code using safe functions/methods like `AsciiExt::make_ascii_lowercase` that use unsafe code (such as `transmute::<&mut str, &mut [u8]>`) internally. It’s up to these functions/methods to preserve the invariant in the safe abstraction they provide.

**Comment by SimonSapin** [comment]
I’ll rebase once a decision is made on what to include or not.

File: src/libcollections/str.rs [hunk 1]
```diff
@@ -806,6 +814,12 @@ impl str {
         core_str::StrExt::split_at(self, mid)
     }
 
+    /// Divide one mutable string slice into two at an index.
+    #[inline]
+    pub fn split_at_mut(&mut self, mid: usize) -> (&mut str, &mut str) {
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: Can you tag this with an explicit `unstable` tag?

File: src/libcollections/str.rs [hunk 2]
```diff
@@ -567,6 +567,14 @@ impl str {
         core_str::StrExt::slice_unchecked(self, begin, end)
     }
 
+    /// Takes a bytewise mutable slice from a string.
+    ///
+    /// Same as `slice_unchecked`, but works with `&mut str` instead of `&str`.
+    #[stable(feature = "derefmut_for_string", since = "1.2.0")]
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: Can this enter as unstable?

**Comment by alexcrichton** [comment]
The consensus of the libs team is to move ahead with this PR, with a rebase and a few nits addressed this should be good to go, thanks @SimonSapin!

**Comment by SimonSapin** [comment]
Rebased and added a commit to address comments. This commits also adds `#[unstable]` to `str::split_at`, which existed before this PR but did not have any stability marker.

I’m getting an LLVM assertion error when building locally :( Could this be specific to stage1? I haven’t tried stage2 yet.

```
../../projects/rust/src/libstd/lib.rs:1:1: 1:1 note: expansion site
rustc: ../../../../../../../projects/rust/src/llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp:525: bool isUnsignedDIType(llvm::DwarfDebug*, const llvm::DIType*): Assertion `T == dwarf::DW_TAG_typedef || T == dwarf::DW_TAG_const_type || T == dwarf::DW_TAG_volatile_type || T == dwarf::DW_TAG_restrict_type || T == dwarf::DW_TAG_enumeration_type' failed.
/home/simon/projects/rust/mk/tests.mk:394: recipe for target 'x86_64-unknown-linux-gnu/stage1/test/stdtest-x86_64-unknown-linux-gnu' failed
make: *** [x86_64-unknown-linux-gnu/stage1/test/stdtest-x86_64-unknown-linux-gnu] Aborted (core dumped)

**Comment by SimonSapin** [comment]
Same error for `make check-stage2-std` :(

**Comment by alexcrichton** [comment]
Unfortunately I don't quite know what's going on there, but it may be either an LLVM bug or a trans bug on our end. If you can emit IR then you can use `bugpoint` to possibly reduce it, otherwise it may need to be reduced manually if it's a compiler bug.

**Comment by SimonSapin** [comment]
Ok, I managed to emit IR by copying a large command from `make VERBOSE=1`, removing `-o ...`, and adding `--emit=llvm-ir`. I still get the assertion error with ``--emit=obj` instead. How can I use bugpoint?

**Comment by SimonSapin** [comment]
Running llc directly:
```
llc: ../../../../../../../projects/rust/src/llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp:525: bool isUnsignedDIType(llvm::DwarfDebug*, const llvm::DIType*): Assertion `T == dwarf::DW_TAG_typedef || T == dwarf::DW_TAG_const_type || T == dwarf::DW_TAG_volatile_type || T == dwarf::DW_TAG_restrict_type || T == dwarf::DW_TAG_enumeration_type' failed.
0  llc             0x0000000001381905 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  llc             0x0000000001380471
2  libpthread.so.0 0x00007f01eb9a9660
3  libc.so.6       0x00007f01ea94d528 gsignal + 56
4  libc.so.6       0x00007f01ea94e93a abort + 362
5  libc.so.6       0x00007f01ea9463a7
6  libc.so.6       0x00007f01ea946452
7  llc             0x0000000000cd601b
8  llc             0x0000000000cd61b2 llvm::DwarfUnit::addConstantValue(llvm::DIE&, llvm::MachineOperand const&, llvm::DIType const*) + 34
9  llc             0x0000000000d013d8 llvm::DwarfCompileUnit::constructVariableDIEImpl(llvm::DbgVariable const&, bool) + 504
10 llc             0x0000000000d01784 llvm::DwarfCompileUnit::constructVariableDIE(llvm::DbgVariable&, bool) + 20
11 llc             0x0000000000d017b6 llvm::DwarfCompileUnit::constructVariableDIE(llvm::DbgVariable&, llvm::LexicalScope const&, llvm::DIE*&) + 22
12 llc             0x0000000000d05c6a llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 330
13 llc             0x0000000000d052d6 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 374
14 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
15 llc             0x0000000000d056c3 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 1379
16 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
17 llc             0x0000000000d052d6 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 374
18 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
19 llc             0x0000000000d056c3 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 1379
20 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
21 llc             0x0000000000d052d6 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 374
22 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
23 llc             0x0000000000d056c3 llvm::DwarfCompileUnit::constructScopeDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&) + 1379
24 llc             0x0000000000d05d53 llvm::DwarfCompileUnit::createScopeChildrenDIE(llvm::LexicalScope*, llvm::SmallVectorImpl<std::unique_ptr<llvm::DIE, std::default_delete<llvm::DIE> > >&, unsigned int*) + 563
25 llc             0x0000000000d05e83 llvm::DwarfCompileUnit::createAndAddScopeChildren(llvm::LexicalScope*, llvm::DIE&) + 67
26 llc             0x0000000000d06188 llvm::DwarfCompileUnit::constructSubprogramScopeDIE(llvm::LexicalScope*) + 280
27 llc             0x0000000000ccdf5f llvm::DwarfDebug::endFunction(llvm::MachineFunction const*) + 1247
28 llc             0x0000000000cb45b8 llvm::AsmPrinter::EmitFunctionBody() + 5160
29 llc             0x00000000005f3136
30 llc             0x0000000001259958 llvm::FPPassManager::runOnFunction(llvm::Function&) + 632
31 llc             0x0000000001259d3b llvm::FPPassManager::runOnModule(llvm::Module&) + 43
32 llc             0x000000000125948f llvm::legacy::PassManagerImpl::run(llvm::Module&) + 783
33 llc             0x00000000005a61d8
34 llc             0x0000000000577a20 main + 352
35 libc.so.6       0x00007f01ea93a790 __libc_start_main + 240
36 llc             0x000000000059fc59 _start + 41
Stack dump:
0.	Program arguments: ./x86_64-unknown-linux-gnu/llvm/Release+Asserts/bin/llc std.ll 
1.	Running pass 'Function Pass Manager' on module 'std.ll'.
2.	Running pass 'X86 Assembly / Object Emitter' on function '@"_ZN5boxed16F.FnBox$LT$A$GT$8call_box20h2184742159468874156E"'
zsh: abort (core dumped)  ./x86_64-unknown-linux-gnu/llvm/Release+Asserts/bin/llc std.ll

**Comment by SimonSapin** [comment]
bugpoint+llvm-dis produced a `bugpoint-reduced-simplified.ll` file that’s 18 MB, which doesn’t look very reduced :/

**Comment by SimonSapin** [comment]
I’m still getting the same LLVM error after rebasing on today’s Rust master. Any suggestion what to do here?

**Comment by alexcrichton** [comment]
Unfortunately I don't know how to get past that error.

**Comment by pnkfelix** [comment]
This seems like it has been blocked for 6 days due to #26484.  However this is unnecessary, assuming it does not add any *new* breakage to `--enable-debug --enable-optimize`, since we have no bot that tests that configuration (see also #27010).

**Comment by pnkfelix** [comment]
After discussion with @SimonSapin I am interpreting this statement from @alexcrichton :+1: 

> The consensus of the libs team is to move ahead with this PR, with a rebase and a few nits addressed this should be good to go, thanks @SimonSapin!

as an implicit `r=me` for the PR.

**Comment by SimonSapin** [comment]
doener on IRC suggested that I might be hitting #26484. After running `configure` again with default flags (I’m not sure what I had before), `make check-stage1-std` succeeded locally.

**Comment by SimonSapin** [comment]
I just pushed a rebase.

**Comment by SimonSapin** [comment]
@alexcrichton With that last commit `make check-stage1` (not just `check-stage1-std`) passes locally.

**Pull request merged**